src/systems/variational_smoother_system.C

```diff
@@ -0,0 +1,467 @@
+      // variable defined above. This is also the Jacobian the 'request_jacobian'
+      // variable refers to. The second mesning refers to the Jacobian of the
+      // physical-to-reference element mapping. This is the Jacobian used to
+      // compute the distorion-dilation metric.
+      //
+      // Grab the physical-to-reference mapping Jacobian matrix (i.e., "S") at this qp
+      RealTensor S;
+      // RealTensors are always 3x3, so we will fill any dimensions above dim
+      // with 1s on the diagonal. This indicates a 1 to 1 relationship between
+      // the physical and reference elements in these extra dimensions.
+      switch (dim)
+      {
+        case 1:
+          S = RealTensor(
+            dxyzdxi[qp](0), 0, 0,
+                         0, 1, 0,
+                         0, 0, 1
+          );
+          break;
+
+        case 2:
+          S = RealTensor(
+            dxyzdxi[qp](0), dxyzdeta[qp](0), 0,
+            dxyzdxi[qp](1), dxyzdeta[qp](1), 0,
+            0,              0,               1
+          );
+          break;
+
+        case 3:
+          S = RealTensor(
+            dxyzdxi[qp],
+            dxyzdeta[qp],
+            dxyzdzeta[qp]
+          ).transpose();  // Note the transposition!
+          break;
+
+        default:
+          libmesh_error_msg("Unsupported dimension.");
+      }
+
+      // Compute quantities needed for the smoothing algorithm
+
+      // determinant
+      const Real det = S.det();
+      const Real det_sq = det * det;
+      const Real det_cube = det_sq * det;
+
+      const Real ref_vol_sq = _ref_vol * _ref_vol;
+
+      // trace of S^T * S
+      // DO NOT USE RealTensor.tr for the trace, it will NOT be correct for
+      // 1D and 2D meshes because of our hack of putting 1s in the diagonal of
+      // S for the extra dimensions
+      const RealTensor ST_S = S.transpose() * S;
+      Real tr = 0.0;
+      for (const auto i : make_range(dim))
+        tr += ST_S(i, i);
+      const Real tr_div_dim = tr / dim;
+
+      // inverse of S
+      const RealTensor S_inv = S.inverse();
+      // inverse transpose of S
+      const RealTensor S_inv_T = S_inv.transpose();
+
+      // Identity matrix
+      const RealTensor I(1, 0, 0,
+                         0, 1, 0,
+                         0, 0, 1);
+
+      // The chi function allows us to handle degenerate elements
+      // When the element is not degenerate (i.e., det(S) > 0), we can set
+      // _epsilon_squared to zero to get chi(det(S)) = det(S)
+      const Real epsilon_sq = det > 0. ? 0. : _epsilon_squared;
+      const Real chi = 0.5 * (det + std::sqrt(epsilon_sq + det_sq));
+      const Real chi_sq = chi * chi;
+      const Real chi_cube = chi_sq * chi;
+      const Real sqrt_term = std::sqrt(epsilon_sq + det_sq);
+      // dchi(x) / dx
+      const Real chi_prime = 0.5 * (1. + det / sqrt_term);
+      // dchi(det(S) / dS
+      RealTensor dchi_dS = chi_prime * det * S_inv_T;
```

Comment:
This could be `const` too, right?  Likewise with `dmu_dS` below?

Normally when I see something that isn't const but could be I figure the programmer was just being lazy, but you're being so careful above that it makes me wonder if I'm missing a subtlety here.

Reply:
Nah, I just missed the const. Good catch!